Sprint 8 solution

Naz1anmak opened the pull request:
- Добавлен список задач в порядке приоритета
  - Задачи и подзадачи теперь нужно создавать со временем начала выполнения и продолжительностью выполнения.
  - Изменены методы в InMemoryTaskManager под новую функциональность.
- Создан абстрактный класс TaskManagerTest для тестов классов от интерфейса TaskManager
- Изменены тесты в InMemoryHistoryManagerTest под новую функциональность.

Naz1anmak committed:
Version 5.0
- Добавлен список задач в порядке приоритета
  - Задачи и подзадачи теперь нужно создавать со временем начала выполнения и продолжительностью выполнения.
  - Изменены методы в InMemoryTaskManager под новую функциональность.
- Создан абстрактный класс TaskManagerTest для тестов классов от интерфейса TaskManager
- Изменены тесты в InMemoryHistoryManagerTest под новую функциональность.

afomkina commented on src/history/HistoryManager.java:
В менеджере историй изменений быть не должно

afomkina commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -11,9 +11,12 @@
 import java.util.List;
 
 public class FileBackedTaskManager extends InMemoryTaskManager {
+    private static InMemoryHistoryManager historyManager;
```
Это поле нужно удалить

afomkina commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -49,6 +52,7 @@ public static FileBackedTaskManager loadFromFile(File file) {
         }
 
         FileBackedTaskManager fileBackedTaskManager = new FileBackedTaskManager(file);
+        historyManager = fileBackedTaskManager.getHistoryManager();
```
Вызов истории тут не нужен

afomkina commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -4,12 +4,15 @@
+import java.time.LocalDateTime;
 import java.util.*;
 
 public class InMemoryHistoryManager implements HistoryManager {
```
Вся логика по сортировке задач должна быть в менеджере задач

afomkina commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -243,33 +295,73 @@ public void updateSub(Subtask oldSubtask, Subtask newSubtask) {
+
+    @Override
+    public List<Task> getPrioritizedTasks() {
+        return historyManager.getPrioritizedTasks();
```
Здесь должна быть реализация, а не вызов метода из истории задач

afomkina commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -243,33 +295,73 @@ public void updateSub(Subtask oldSubtask, Subtask newSubtask) {
+        return historyManager.getPrioritizedTasks();
+    }
+
+    private void getIsIntersectionTasks(Task newTask) {
```
Этот метод нужно удалить

afomkina commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -243,33 +295,73 @@ public void updateSub(Subtask oldSubtask, Subtask newSubtask) {
+        }
+    }
+
+    private void getIsIntersectionTasks(Task updTask, int idAddedTask) {
```
Метод должен возвращать `boolean` и реализовывать поиск пересечений, а не вызывать историю задач

Naz1anmak committed:
Version 5.1
- Логика работы с приоритетными задачами перенесена в InMemoryTaskManager

afomkina commented on src/manager/TaskManager.java:
```diff
@@ -1,31 +1,32 @@
     List<Task> getTasks();
 
-    void updateTask(Task task);
+    void updateTask(Task oldTask, Task newTask);
```
В методах обновления задач изменений быть не должно. В параметрах метода всегда должно передаваться задача для обновления.
Нужно убрать изменения из всех методов обновления задач по типам

afomkina commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -243,33 +305,93 @@ public void updateSub(Subtask oldSubtask, Subtask newSubtask) {
+        return List.copyOf(prioritizedTasks);
+    }
+
+    public boolean isIntersectionTasks(Task newTask) {
```
Этот метод и метод ниже отличаются одним параметром. Весь остальной код дублируется. Нужно объединить два метода в один и сделать дополнительную фильтрацию в зависимости от значения параметра `idAddedTask`.
Метод должен быть приватным, так как используется только в данном классе

afomkina commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -243,33 +305,93 @@ public void updateSub(Subtask oldSubtask, Subtask newSubtask) {
+        }
+    }
+
+    public void removePrioritizedTask(Task task) {
```
Методы со строк 343 и 349 должны быть приватными

afomkina commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -71,23 +97,26 @@ public int addNewEpic(Epic epic) {
+        updateEpicStatus(epicCopy);
+        epicCopy.setId(epic.getId());
+        epicCopy.setDuration(epic.getDuration());
+        epicCopy.setStartTime(epic.getStartTime());
```
Метод получения эпика по id должен возвращать нужный эпик и записывать данные в менеджер историю. Другая логика - лишняя, ее нужно удалить

afomkina commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -1,43 +1,55 @@
+
         int taskId = idCounter++;
-        Task taskCopy = new Task(task.getName(), task.getDescription(), task.getStatus());
+        Task taskCopy = new Task(task.getName(), task.getDescription(),
```
Создавать копию задачи не нужно

afomkina commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -71,23 +97,26 @@ public int addNewEpic(Epic epic) {
         Epic epicCopy = new Epic(epic.getName(), epic.getDescription());
```
Создавать копию эпика не нужно

afomkina commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -96,21 +125,29 @@ public List<Epic> getEpics() {
+        newEpic.setId(oldEpic.getId());
+        if (!oldEpic.getSubtaskIds().isEmpty()) {
+            oldEpic.getSubtaskIds().forEach(newEpic::addSubtaskId);
+            updateEpicStatus(newEpic);
```
В методе обновления эпика, пересчитывать его статус не нужно

afomkina commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -96,21 +125,29 @@ public List<Epic> getEpics() {
     }
 
     @Override
     public void updateEpicStatus(Epic epic) {
```
Метод должен быть приватным, из интерфейса его нужно убрать. 
Подсчет времени нужно вынести в отдельный метод

Naz1anmak replied:
Метод используется в FileBackedTaskManager, сделал его protected

afomkina commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -170,44 +222,48 @@ public int addNewSubtask(Subtask subtask) {
 
         int subtaskId = idCounter++;
         Subtask subtaskCopy =
```
Создавать копию подзадачи не нужно

Naz1anmak committed:
Version 5.2
- Изменены области видимости некоторых методов.
- Добавлены новые конструкторы.
- Классы Task, Epic и Subtask теперь частично иммутабельные.
